### PR TITLE
Add CategoricalColorMapper

### PR DESCRIPTION
**Pull request by birdsarah**
Fixes: #5013

See #5090 for discussion.

### CODE REVIEW

**Comment by birdsarah** [comment]
@bryevdv - merge when green?

**Comment by bryevdv** [comment]
There's a real unit test failure (defaults mismatch)

**Comment by birdsarah** [comment]
Why didn't it fail before?

**Comment by birdsarah** [comment]
I've fixed the defaults mismatch. But it seems very odd to me that it didn't fail before but the mismatch still existed (default on bokehjs side, but not on python side).

**Comment by bryevdv** [comment]
Yah it's weird, the JS tests really need a pass over in a few different ways.

File: bokehjs/src/coffee/models/mappers/categorical_color_mapper.coffee [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+  _get_values: (data, palette) ->
+    values = []
+
+    for d in data
+      key = @factors.indexOf(d)
+      if key < 0
+        color = @nan_color
+      else
+        color = palette[key]
```
Comment (bryevdv) [inline bokehjs/src/coffee/models/mappers/categorical_color_mapper.coffee]: What if the palette is too small?

Comment (birdsarah) [inline bokehjs/src/coffee/models/mappers/categorical_color_mapper.coffee]: It errors miserably :)

File: bokeh/models/mappers.py [hunk 2]
```diff
@@ -24,6 +25,45 @@ class ColorMapper(Model):
...
+
+    def __init__(self, **kwargs):
+        super(ColorMapper, self).__init__(**kwargs)
+        palette = self.palette
+        factors = self.factors
+        if palette and factors:
+            if len(palette) != len(factors):
+                extra_factors = factors[len(palette):]
+                warnings.warn("""Palette length does not match number of
+factors. %s will be assigned to `nan_color` %s""" % (extra_factors, self.nan_color))
```
Comment (bryevdv) [inline bokeh/models/mappers.py]: This seems like an odd message if the palette is too long. `[] will be assigned to 'nan_color'`

Comment (birdsarah) [inline bokeh/models/mappers.py]: Agreed, late night coding sigh.

**Pull request merged**